derive/src/parse.rs

```diff
@@ -196,7 +196,7 @@ pub fn next_group(source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Opt
 //     println!("{:?}", source.peek());
 // }
 
-fn next_type<T: Iterator<Item = TokenTree>>(mut source: &mut Peekable<T>, generic_typenames: &HashMap<String, String> ) -> Option<Type> {
+fn next_type<T: Iterator<Item = TokenTree>>(mut source: &mut Peekable<T>, generic_typenames: &HashMap<&String, &Vec<String>> ) -> Option<Type> {
```

Comment:
I know this is actually a part of previous PR, but could you reiterate why we need to provide additional context to `next_type` combinator? it looks sort of against the very idea of parser combinators - each combinator gets the next token or None if it fails to get this token  

in other words, all the other `next_whatever` functions are way simpler and do not have any additional context

asking because with this change:
```diff
diff --git a/derive/src/parse.rs b/derive/src/parse.rs
index e1a5bee..d03a352 100644
--- a/derive/src/parse.rs
+++ b/derive/src/parse.rs
@@ -12,14 +12,12 @@ use alloc::{format, vec};
 
 use proc_macro::{Delimiter, Group, TokenStream, TokenTree};
 
-
 #[cfg(features = "no_std")]
 use hashbrown::HashMap;
 
 #[cfg(not(features = "no_std"))]
 use std::collections::HashMap;
 
-
 #[derive(Debug)]
 pub struct Attribute {
     pub name: String,
@@ -56,7 +54,7 @@ pub struct Struct {
     pub named: bool,
     pub fields: Vec<Field>,
     pub attributes: Vec<Attribute>,
-    pub generics: HashMap<String, String>
+    pub generics: HashMap<String, String>,
 }
 
 #[derive(Debug)]
@@ -72,7 +70,7 @@ pub struct Enum {
     pub name: String,
     pub variants: Vec<EnumVariant>,
     pub attributes: Vec<Attribute>,
-    pub generics: HashMap<String, String>
+    pub generics: HashMap<String, String>,
 }
 
 #[allow(dead_code)]
@@ -196,7 +194,7 @@ pub fn next_group(source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Opt
 //     println!("{:?}", source.peek());
 // }
 
-fn next_type<T: Iterator<Item = TokenTree>>(mut source: &mut Peekable<T>, generic_typenames: &HashMap<String, String> ) -> Option<Type> {
+fn next_type<T: Iterator<Item = TokenTree>>(mut source: &mut Peekable<T>) -> Option<Type> {
     let group = next_group(&mut source);
     if group.is_some() {
         let mut group = group.unwrap().stream().into_iter().peekable();
@@ -206,7 +204,7 @@ fn next_type<T: Iterator<Item = TokenTree>>(mut source: &mut Peekable<T>, generi
             path: "".to_string(),
         };
 
-        while let Some(next_ty) = next_type(&mut group, generic_typenames) {
+        while let Some(next_ty) = next_type(&mut group) {
             tuple_type.path.push_str(&format!("{}, ", next_ty.path));
         }
 
@@ -224,9 +222,9 @@ fn next_type<T: Iterator<Item = TokenTree>>(mut source: &mut Peekable<T>, generi
 
     let angel_bracket = next_exact_punct(&mut source, "<");
     if angel_bracket.is_some() {
-        let mut generic_type = next_type(source, generic_typenames).expect("Expecting generic argument");
+        let mut generic_type = next_type(source).expect("Expecting generic argument");
         while let Some(_comma) = next_exact_punct(&mut source, ",") {
-            let next_ty = next_type(source, generic_typenames).expect("Expecting generic argument");
+            let next_ty = next_type(source).expect("Expecting generic argument");
             generic_type.path.push_str(&format!(", {}", next_ty.path));
         }
 
@@ -321,7 +319,6 @@ fn next_attributes_list(source: &mut Peekable<impl Iterator<Item = TokenTree>>)
 fn next_fields(
     mut body: &mut Peekable<impl Iterator<Item = TokenTree>>,
     named: bool,
-    generic_typenames: &HashMap<String, String>
 ) -> Vec<Field> {
     let mut fields = vec![];
 
@@ -341,7 +338,7 @@ fn next_fields(
             None
         };
 
-        let ty = next_type(&mut body, &generic_typenames).expect("Expected field type");
+        let ty = next_type(&mut body).expect("Expected field type");
         let _punct = next_punct(&mut body);
 
         fields.push(Field {
@@ -371,7 +368,7 @@ fn next_struct(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> St
             generics: Default::default(),
         };
     };
-    
+
     let group = group.unwrap();
     let delimiter = group.delimiter();
     let named = match delimiter {
@@ -382,7 +379,7 @@ fn next_struct(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> St
     };
 
     let mut body = group.stream().into_iter().peekable();
-    let fields = next_fields(&mut body, named, &generic_types);
+    let fields = next_fields(&mut body, named);
 
     if named == false {
         next_exact_punct(&mut source, ";").expect("Expected ; on the end of tuple struct");
@@ -393,7 +390,7 @@ fn next_struct(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> St
         named,
         fields,
         attributes: vec![],
-        generics: generic_types
+        generics: generic_types,
     }
 }
 
@@ -407,7 +404,8 @@ fn next_enum(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Enum
             name: enum_name,
             variants: vec![],
             attributes: vec![],
-            generics: Default::default(),        };
+            generics: Default::default(),
+        };
     };
     let group = group.unwrap();
     let mut body = group.stream().into_iter().peekable();
@@ -442,7 +440,7 @@ fn next_enum(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Enum
         };
         {
             let mut body = group.stream().into_iter().peekable();
-            let fields = next_fields(&mut body, named, &generic_types);
+            let fields = next_fields(&mut body, named);
             variants.push(EnumVariant {
                 name: variant_name,
                 named,
@@ -458,7 +456,7 @@ fn next_enum(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Enum
         name: enum_name,
         variants,
         attributes: vec![],
-        generics: generic_types
+        generics: generic_types,
     }
 }
 
@@ -468,7 +466,7 @@ fn get_bounds(source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> HashMap
     match source.peek() {
         Some(content) if content.to_string() == "<" => {
             source.next();
-            let mut typename = "".to_string(); 
+            let mut typename = "".to_string();
             let mut generic_bound = String::new();
             let mut in_type = true;
             while let Some(tok) = source.next() {
@@ -477,32 +475,32 @@ fn get_bounds(source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> HashMap
                         ret.insert(typename.clone(), generic_bound.clone());
                         typename.clear();
                         generic_bound.clear();
-                        break
-                    },
+                        break;
+                    }
                     ":" => {
                         if in_type {
                             in_type = false
                         } else {
                             generic_bound += ":";
                         }
-                    },
+                    }
                     "," => {
                         ret.insert(typename.clone(), generic_bound.clone());
                         typename.clear();
                         generic_bound.clear();
                         in_type = true
-                    },
+                    }
                     c => {
                         if in_type {
-                            typename+=c;
+                            typename += c;
                         } else {
                             generic_bound += c;
                         }
                     }
                 }
             }
-        },
-        _ => ()
+        }
+        _ => (),
     }
     ret
 }
```

all the tests still pass?

Reply:
I've never really worked with tokenizers before, so idk what best practice is here. That definitely looks simpler though, will mess with it and make sure the tests are good enough to trust with respect to that.

Reply:
Yup, that worked just fine. Not sure what I had in mind back then, but all gone now